---
title: "Assembly result"
output:
  html_document:
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
---

```{r setup, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.align="center")
library(DT)
library(dplyr)
```

***

> __File path: 1_Assembly/__

#### {Assembly from PacBio subreads}

Flye, a long-read assembly algorithm that generates arbitrary paths in an unknown repeat graph, called disjointigs, and constructs an accurate repeat graph from these error-riddled disjointigs.
[doi:10.1038/s41587-019-0072-8] [1]

For more detail of each step of assembly algorithms, please view the introduction:

<a href="https://github.com/fenderglass/Flye/blob/flye/docs/USAGE.md" target="_blank">flye detail introduction</a>

#### {Scaffolding & Polishing}

1. Scaffolding by SSPACE-Long

Boetzer, M., Pirovano, W. SSPACE-LongRead: scaffolding bacterial draft genomes using long read sequence information. BMC Bioinformatics 15, 211 (2014) doi:10.1186/1471-2105-15-211

2. Polishing by Arrow

The GenomicConsensus package provides the variantCaller tool, which allows you to apply the Arrow algorithm to mapped PacBio reads to get consensus and variant calls.

https://github.com/PacificBiosciences/GenomicConsensus

#### {Fixing & Circularization}

Current long-read assembly software still typically assumes that the contigs they produce are linear. 

In contrast, the genome of almost every species contains at least one circular DNA structure, such as bacterial chromosomes and plasmids. 

Thus, a useful circularization of genome assembly tool – “Circlator v1.5.5” is our default pipeline to correct and linearize the genome [2]. 

#### {Assembly evaluation}

Finally, we are using “QUAST v4.5” for evaluating the assembled genome quality [3].

***

### Flye + SSPACE-Long + Arrow assembly summary

[Flye + SSPACE-Long + Arrow assembled scaffolds (.fasta)](files/1_Assembly/<prefix>.scaffolds.polished.fasta)

a.) Contigs were assembled by Flye.

b.) Scaffolds were generated by SSPACE-Long

c.) Polished scaffolds were polished by Arrow

***

### Fixing and Circularising polished scaffolds by Circlator

<a href="https://sanger-pathogens.github.io/circlator/" target="_blank">Circlator is a tool to circularize genome</a>

a.) Identify each circular sequence and output a linearised version of it.

b.) Join contigs together.

c.) If a dnaA gene is found, then that is used as the starting position.

[Fixed and circularised contigs (.fasta)](files/1_Assembly/<prefix>_polished_fixed.circularise.fasta) <--- used for following analysis

<a href="files/1_Assembly/quast/report.html" target="_blank">[Assembly Report by QUAST (.html)]</a> <--- assembly evaluation

```{r, echo=FALSE, message=FALSE, warning=FALSE}
fixed_polished_report <- read.table("files/1_Assembly/quast/report.tsv", header=T, check.names=F, sep="\t")

datatable(fixed_polished_report, options = list(
  dom = 't',
  columnDefs = list(list(className = 'dt-center', targets = c(2,3,4)))
)) %>%
  formatCurrency(c('contigs','scaffolds-polished','circlator'), currency = "", interval = 3, mark = ",", digits=0)
```

***

> __File path: 2_Subread/__

__Each polymerase read is partitioned to form one or more subreads, which contain sequence from a single pass of a polymerase on a single strand of an insert within a SMRTbell™ template and no adapter sequences. The subreads contain the full set of quality values and kinetic measurements. Subreads are useful for applications like de novo assembly, resequencing, base modification analysis, and so on.__

```
[Notice]:
1. Due to subreads fasta/fastq file is too large, we’re not provided in the report. If user required, get from your download link or hard-copy is a better alternative data transfer method.

2. If user are interesting on the PacBio subreads distribution and stats, please view the PDF in your download folder or let us know.
```

***

### Reference
```
1. Flye v2.5: Fast and accurate de novo assembler for single molecule sequencing reads
Mikhail Kolmogorov, Jeffrey Yuan, Yu Lin & Pavel A. Pevzner. (2019) doi:10.1038/s41587-019-0072-8

2. "Circlator v1.4.0" - Assembly fixed & improvement
Hunt M, Silva ND, Otto TD, Parkhill J, Keane JA, Harris SR. Circlator: automated circularization of genome assemblies using long sequencing reads. Genome Biology. 2015;16:294. 
doi:10.1186/s13059-015-0849-0.

3. "QUAST v4.5" - Assembly evalutation
Gurevich A., Saveliev V., Vyahhi N., Tesler G. QUAST: quality assessment tool for genome assemblies. Bioinformatics, 2013.
```
