---
title: "Furthur Annotation"
output:
  html_document:
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.align="center")
library(DT)
library(dplyr)
library(readxl)
library(plotly)
```

***

## Gene Prediction by Transdecoder

```
File path: files/4_Annotation/Gene_prediction
```

<br>

TransDecoder identifies candidate coding regions within transcript sequences, such as those generated by de-novo/ref-based RNA-Seq transcript assembly using Trinity [1].

TransDecoder identifies likely coding sequences based on the following criteria:

- a minimum length open reading frame (ORF) is found in a transcript sequence

- a log-likelihood score similar to what is computed by the GeneID software is > 0.

- the above coding score is greatest when the ORF is scored in the 1st reading frame as compared to scores in the other 2 forward reading frames.

- if a candidate ORF is found fully encapsulated by the coordinates of another candidate ORF, the longer one is reported. However, a single transcript can report multiple ORFs (allowing for operons, chimeras, etc).

- a PSSM is built/trained/used to refine the start codon prediction.

```{r, echo=FALSE, message=FALSE, warning=FALSE}
df <- read.delim("files/4_Annotation/Gene_prediction/README.txt", header=T, check.names=F, sep="\t")
df$Extension <- paste0("[",df$Extension,"]","(files/4_Annotation/Gene_prediction/",df$Extension,")")
knitr::kable(df)
```

***

## BLAST Annotation (database: uniprot-swissprot) {.tabset .tabset-fade}

```
File path: files/4_Annotation/blast_result
```

<br>

### BLASTx

- Transcript.95.fasta (nucleotide) -> Protein DB

[Full Table Download (.xlsx)](files/4_Annotation/blast_result/uniprot.blastx.report.xlsx)

```{r, echo=FALSE, message=FALSE, warning=FALSE}
nucl_blast <- read_excel("files/4_Annotation/blast_result/uniprot.blastx.report.xlsx")
datatable(nucl_blast,
    escape=F,
    filter = "top",
    rownames= F,
    extensions = list("ColReorder" = NULL,
                      "Buttons" = NULL),
    options = list(
                dom = 'BRrltpi',
                autoWidth=TRUE,
                ColReorder = TRUE,
                scrollX = TRUE,
                #columnDefs = list(
                #  list(className = "dt-center", targets = c(2:11,14:17)),
                #  list(targets = c(3:12,13:14), visible = FALSE)
                #),
                buttons =
                  list(
                    'copy',
                    'print',
                    list(
                      extend = 'collection',
                      buttons = c('csv', 'excel', 'pdf'),
                      text = 'Download'
                    ),
                    I('colvis')
                  )
              )
)
```

<br>

### BLASTp

- Transcript.95.fasta.transdecoder.pep (protein) -> Protein DB

[Full Table Download (.xlsx)](files/4_Annotation/blast_result/uniprot.blastp.report.xlsx)

```{r, echo=FALSE, message=FALSE, warning=FALSE}
prot_blast <- read_excel("files/4_Annotation/blast_result/uniprot.blastp.report.xlsx")
datatable(prot_blast,
    escape=F,
    filter = "top",
    rownames= F,
    extensions = list("ColReorder" = NULL,
                      "Buttons" = NULL),
    options = list(
                dom = 'BRrltpi',
                autoWidth=TRUE,
                ColReorder = TRUE,
                scrollX = TRUE,
                #columnDefs = list(
                #  list(className = "dt-center", targets = c(2:11,14:17)),
                #  list(targets = c(3:12,13:14), visible = FALSE)
                #),
                buttons =
                  list(
                    'copy',
                    'print',
                    list(
                      extend = 'collection',
                      buttons = c('csv', 'excel', 'pdf'),
                      text = 'Download'
                    ),
                    I('colvis')
                  )
              )
)
```

***

## Protein Group Orthologue (COG)

```
File path: files/4_Annotation/COG
```

<br>

In order to extract the maximum amount of information from the rapidly accumulating genome sequences, all conserved genes need to be classified according to their homologous relationships. Each COG consists of individual orthologous proteins or orthologous sets of paralogs from at least three lineages. Orthologs typically have the same function, allowing transfer of functional information from one member to an entire COG.

[Full Table Download (.xlsx)](files/4_Annotation/COG/rps-blast_cog.txt.xlsx)

```{r, echo=FALSE, message=FALSE, warning=FALSE}
cog_report <- read_excel("files/4_Annotation/COG/rps-blast_cog.txt.xlsx")
datatable(cog_report,
    escape=F,
    filter = "top",
    rownames= F,
    extensions = list("ColReorder" = NULL,
                      "Buttons" = NULL),
    options = list(
                dom = 'BRrltpi',
                autoWidth=TRUE,
                ColReorder = TRUE,
                scrollX = TRUE,
                columnDefs = list(
                  list(className = "dt-center", targets = c(2:8,9:13)),
                  list(targets = c(3:9, 11, 14:17), visible = FALSE)
                ),
                buttons =
                  list(
                    'copy',
                    'print',
                    list(
                      extend = 'collection',
                      buttons = c('csv', 'excel', 'pdf'),
                      text = 'Download'
                    ),
                    I('colvis')
                  )
              )
)
```

<br>

```{r, out.width = "800px", echo=FALSE, message=FALSE, warning=FALSE}
knitr::include_graphics("files/4_Annotation/COG/cog_barchart.png")
```

***

## Gene Ontology (GO)

```
File path: files/4_Annotation/GO
```

<br>

### General Annotation Summary

Gene ontology concern with annotation of genes and gene products and to provide centralized access to resources and tools. both GO and COG provide specific information about gene or gene products.

__There are three main classes in GO database:__

__1. Cellular Component: These terms describe a component of a cell that is part of a larger object, such as an anatomical structure (e.g. rough endoplasmic reticulum or nucleus) or a gene product group (e.g. ribosome, proteasome or a protein dimer).__

__2.Biological Process: A biological process term describes a series of events accomplished by one or more organized assemblies of molecular functions.__

__3. Molecular Function: Molecular function terms describes activities that occur at the molecular level, such as "catalytic activity" or "binding activity".__

<br>

[Full Table Download (.xlsx)](files/4_Annotation/GO/GO_mapping.xlsx)

```{r, echo=FALSE, message=FALSE, warning=FALSE}
go_report <- read.delim("files/4_Annotation/GO/GO_mapping.txt", header=T, check.names=F, sep="\t")
datatable(go_report,
    escape=F,
    filter = "top",
    rownames= F,
    extensions = list("ColReorder" = NULL,
                      "Buttons" = NULL),
    options = list(
                dom = 'BRrltpi',
                autoWidth=TRUE,
                ColReorder = TRUE,
                scrollX = TRUE,
                columnDefs = list(
                  list(className = "dt-center", targets = 3),
                  #list(targets = c(3:9, 11, 14:17), visible = FALSE)
                  list(targets=4, 
                       render = JS("function(data, type, row, meta) {",
                                   "return type === 'display' && data.length > 20 ?",
                                   "'<span title=\"' + data + '\">' + data.substr(0, 20) + '...</span>' : data;","}"))
                ),
                buttons =
                  list(
                    'copy',
                    'print',
                    list(
                      extend = 'collection',
                      buttons = c('csv', 'excel', 'pdf'),
                      text = 'Download'
                    ),
                    I('colvis')
                  )
              )
)
```

<br>

```{r, out.width = "800px", echo=FALSE, message=FALSE, warning=FALSE}
knitr::include_graphics("files/4_Annotation/GO/GO_barchart.png")
```

<br>

### GO enrichment by UP/DOWN regulation (according to comparisons) {.tabset .tabset-fade}

<br>

<a href="http://revigo.irb.hr/" target="_blank">REVIGO: summarizes and visualizes long lists of Gene Ontology terms</a> [3]

###GO_COMPARISON###

***

## KEGG pathway (KEGG)

```
File path: files/4_Annotation/KEGG
```

<br>

KEGG PATHWAY is a collection of manually drawn pathway maps representing our knowledge on the molecular interaction, reaction and relation networks for:

__1. Metabolism: Global/overview, Carbohydrate, Energy, Lipid, Nucleotide, Amino acid, Other amino, Glycan, Cofactor/vitamin, Terpenoid/PK, Other secondary metabolite, Xenobiotics, Chemical structure__

__2. Genetic Information Processing__

__3. Environmental Information Processing__

__4. Cellular Processes__

__5. Organismal Systems__

__6. Human Diseases__

__7. Drug Development__


***

[Full Table Download (.xlsx)](files/4_Annotation/KEGG/ec2kegg.xls)

```
If you would like to get the pathway images, please click the "Column visibility: -> "URL" or downlaod the excel table 
```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
kegg_report <- read_excel("files/4_Annotation/KEGG/ec2kegg.xls")
datatable(kegg_report,
    escape=F,
    filter = "top",
    rownames= F,
    extensions = list("ColReorder" = NULL,
                      "Buttons" = NULL),
    options = list(
                dom = 'BRrltpi',
                autoWidth=TRUE,
                ColReorder = TRUE,
                scrollX = TRUE,
                columnDefs = list(
                  list(className = "dt-center", targets = c(0, 3:8, 15, 16)),
                  list(targets = c(9:11, 12:14, 17), visible = FALSE)
#                  list(targets=17, 
#                       render = JS("function(data, type, row, meta) {",
#                                   "return type === 'display' && data.length > 20 ?",
#                                   "'<span title=\"' + data + '\">' + data.substr(0, 20) + '...</span>' : data;","}"))
                ),
                buttons =
                  list(
                    'copy',
                    'print',
                    list(
                      extend = 'collection',
                      buttons = c('csv', 'excel', 'pdf'),
                      text = 'Download'
                    ),
                    I('colvis')
                  )
              )
)
```

<br>

```{r, out.width = "800px", echo=FALSE, message=FALSE, warning=FALSE}
knitr::include_graphics("files/4_Annotation/KEGG/kegg_barchart.png")
```

<br>

### KEGG pathway annotation by UP/DOWN regulation (according to comparisons) {.tabset .tabset-fade}

###KEGG_COMPARISON###

***

## Final Summary Table

[Final Summary Table](files/final_report.xlsx)

***

## Reference
```
1. Haas BJ, Papanicolaou A, Yassour M, et al. De novo transcript sequence reconstruction from RNA-Seq: reference generation and analysis with Trinity. Nature protocols. 2013;8(8):10.1038/nprot.2013.084. doi:10.1038/nprot.2013.084.

2. https://trinotate.github.io/

3. Supek F, Bošnjak M, Škunca N, Šmuc T. REVIGO Summarizes and Visualizes Long Lists of Gene Ontology Terms. Gibas C, ed. PLoS ONE. 2011;6(7):e21800. doi:10.1371/journal.pone.0021800.
```

***
