---
title: "Assembly result"
output:
  html_document:
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
---

```{r setup, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.align="center")
library(DT)
library(dplyr)
```

***

> __File path: 1_Assembly/__

## WorkFlow of Assembly Pipeline

### Step 1.) Hifi reads Assembly:  "hifiasm v0.8"[1]

Hifiasm is a fast haplotype-resolved de novo assembler for PacBio Hifi reads. It can assemble a human genome in several hours and works with the California redwood genome, one of the most complex genomes sequenced so far.

For more detail of each step of assembly algorithms, please view the introduction:

```{r, echo=FALSE, message=FALSE, warning=FALSE}
info <- read.table("files/1_Assembly/assembly_info.txt", header=T, check.names=F, sep="\t")
datatable(info, options = list(
  dom = 't',
  columnDefs = list(list(className = 'dt-center', targets = c(2,3,4)))
)) %>%
  formatCurrency(c('length','cov.','circ.'), currency = "", interval = 3, mark = ",", digits=0)
```

<a href="https://github.com/chhylp123/hifiasm" target="_blank">hifiasm detail introduction</a>

&nbsp;

### Step 2.) Scaffolding + Gap closing + Polishing


1. Scaffolding by "SSPACE-Long v1.1 [2]" (optional if 1 contig only)

SSPACE-LongRead is a stand-alone program for scaffolding pre-assembled contigs using long reads (e.g. PacBio RS reads). Using the long read information, contigs (or scaffolds) are placed in the right order and orientation in so-called super-scaffolds.

&nbsp;

2. Gap closing by "PBJelly v15.8.24 [3]" (optional if 1 contig only)

PBJelly is software for closing gaps in genome assemblies using long reads [https://doi.org/10.1371/journal.pone.0047768].

&nbsp;

3. Genome polishing by "Arrow v2.3.3" [4]

The GenomicConsensus package provides the variantCaller tool, which allows you to apply the Arrow algorithm to mapped PacBio reads to get consensus and variant calls.

https://github.com/PacificBiosciences/GenomicConsensus

&nbsp;

### Step 3.) FixStart & Circularization: "circlator v1.5.5" [5]

Current long-read assembly software still typically assumes that the contigs they produce are linear. 

In contrast, the genome of almost every species contains at least one circular DNA structure, such as bacterial chromosomes and plasmids. 

Thus, a useful circularization of genome assembly tool – “Circlator v1.5.5” is our default pipeline to correct and linearize the genome. 

&nbsp;

### Step 4.) Assembly evaluation: "QUAST v4.6.3" [6]

Finally, we are using QUAST for evaluating the assembled genome quality.

***

## Final Assembly Summary

&nbsp;

[Flye + SSPACE-Long + Arrow assembled scaffolds (.fasta)](files/1_Assembly/<prefix>.scaffolds.polished.fasta)

a.) Contigs were assembled by Flye.

b.) Scaffolds were generated by SSPACE-Long

c.) Polished scaffolds were polished by Arrow

***

### FixStart and Circularising polished scaffolds by Circlator

a.) Identify each circular sequence and output a linearised version of it.

b.) Join contigs together.

c.) If a dnaA gene is found, then that is used as the starting position.

[FixStart and circularised contigs (.fasta)](files/1_Assembly/<prefix>_polished_fixed.circularise.fasta) <--- final completed genome, used for following analysis

<a href="files/1_Assembly/quast/report.html" target="_blank">[Assembly Report by QUAST (.html)]</a> <--- assembly evaluation

```{r, echo=FALSE, message=FALSE, warning=FALSE}
fixed_polished_report <- read.table("files/1_Assembly/quast/report.tsv", header=T, check.names=F, sep="\t")

datatable(fixed_polished_report, options = list(
  dom = 't',
  columnDefs = list(list(className = 'dt-center', targets = c(2,3,4)))
)) %>%
  formatCurrency(c('contigs','scaffolds-polished','circlator'), currency = "", interval = 3, mark = ",", digits=0)
```

***

> __File path: 2_HIFIreads/__

You Could Found Fastq From Your Account Folder.

```
[What is HiFi reads]:

1. "Highly Accurate" and "Long" Consensus Sequences (at least 3 passes & QV20 accuracy)

2. Produce HiFi reads using the circular consensus sequencing (CCS) mode to provide base-level resolution with >99% single-molecule read accuracy for the detection of all variant types from single nucleotide to structural variants.

3. It might be supported by other 3rd party tools that Illumina could be used.

```

***

### Reference
```
1. hifiasm v0.8: Fast and accurate de novo assembler for single molecule sequencing reads.
https://github.com/chhylp123/hifiasm

2. SSPACE-Long v1.1: SSPACE-LongRead is a stand-alone program for scaffolding pre-assembled contigs using long reads (e.g. PacBio subreads).
Boetzer M, Pirovano W (2014) SSPACE-LongRead: scaffolding bacterial draft genomes using long read sequence information. BMC Bioinformatics, 15:211.

3. PBJelly v15.8.24: PBJelly is software for closing gaps in genome assemblies using long reads 
Adam C. English ,Stephen Richards,Yi Han,Min Wang,Vanesa Vee,Jiaxin Qu,Xiang Qin,Donna M. Muzny,Jeffrey G. Reid,Kim C. Worley,Richard A. Gibbs
doi.org/10.1371/journal.pone.0047768

4. Arrow v2.3.3: The GenomicConsensus package provides the variantCaller tool, which allows you to apply the Arrow algorithm to mapped PacBio reads to get consensus and variant calls.
https://github.com/PacificBiosciences/GenomicConsensus

5. "Circlator v1.5.5" - fix start & improvement draft genome
Hunt M, Silva ND, Otto TD, Parkhill J, Keane JA, Harris SR. Circlator: automated circularization of genome assemblies using long sequencing reads. Genome Biology. 2015;16:294. 
doi:10.1186/s13059-015-0849-0.

6. "QUAST v4.6.3" - Assembly evalutation
Gurevich A., Saveliev V., Vyahhi N., Tesler G. QUAST: quality assessment tool for genome assemblies. Bioinformatics, 2013.
```
